pandas/core/groupby/generic.py

```diff
@@ -591,20 +591,23 @@ def nunique(self, dropna: bool = True) -> Series:
 
         val = self.obj._internal_get_values()
 
-        # GH 27951
-        # temporary fix while we wait for NumPy bug 12629 to be fixed
-        val[isna(val)] = np.datetime64("NaT")
-
-        try:
-            sorter = np.lexsort((val, ids))
-        except TypeError:  # catches object dtypes
-            msg = f"val.dtype must be object, got {val.dtype}"
-            assert val.dtype == object, msg
+        def _object_sorter(val, ids):
```

Comment:
let's just simplifiy this

you can factorize any pandas objects, so just

```
codes, _ = algorithms.factorize(val, sort=False)
sorter = np.lexsort((codes, ids))
mask = codes == -1
```
(I renamed val -> codes), but should just work

Reply:
Thanks @jreback ! Have updated